frontend/src/store/modules/user.ts

```diff
@@ -62,21 +65,39 @@ export default {
       user: firebase.User | null
     ) {
       commit("setCurrentUser", user);
-      if (user) {
-        await dispatch(
-          "updateUserRoles",
-          (await user.getIdTokenResult()).claims.roles
-        );
-      } else {
-        await dispatch("updateUserRoles", null);
-      }
+      dispatch("updateUserRoles", await dispatch("getUserRoles"));
     },
     updateUserRoles(
       { commit, getters }: ActionContext<any, any>,
       roles: { [key: string]: any } | null
     ) {
       commit("setUserRoles", roles);
       getters.ability.update(defineAbilities());
+    },
+    async getUserRoles(
+      { state }: ActionContext<any, any>,
+      refreshToken = false
+    ) {
+      if (!state.currentUser) return null;
+      return (await state.currentUser.getIdTokenResult(refreshToken)).claims
+        .roles;
+    },
+    handleMetaData({ state, dispatch }: ActionContext<any, any>) {
+      if (callback && metadataRef) {
+        metadataRef.off("value", callback);
+      }
+      if (state.currentUser) {
+        callback = async () => {
+          await dispatch("onMetaDataChange");
+        };
+        metadataRef = firebase
+          .database()
+          .ref(`users/${state.currentUser.uid}/refreshTime`);
+        metadataRef.on("value", callback);
```

Comment:
I think we should track the last time when we refreshed the roles and compare it to the new value in the callback. Because “This event will trigger once with the initial data stored at this location, and then trigger again each time the data changes.” (from https://firebase.google.com/docs/reference/js/firebase.database.Reference#value-event). Thus, we will get one callback after hooking even if the time was not changed.

OR

We can use this feature to update the roles after user change. Thus, we don't need to dispatch `updateUserRoles` in `handleUser`.

Consider which approach is better. I like the latter.